manuals/dc.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "DC" "1" "June 2020" "Gavin D. Howard" "General Commands Manual"
.
.SH "NAME"
\fBdc\fR \- arbitrary\-precision reverse\-Polish notation calculator
.
.SH "SYNOPSIS"
\fBdc\fR [\fB\-hiPvVx\fR] [\fB\-\-version\fR] [\fB\-\-help\fR] [\fB\-\-interactive\fR] [\fB\-\-no\-prompt\fR] [\fB\-\-extended\-register\fR] [\fB\-e\fR \fIexpr\fR] [\fB\-\-expression=\fR\fIexpr\fR\.\.\.] [\fB\-f\fR \fIfile\fR\.\.\.] [\fB\-file=\fR\fIfile\fR\.\.\.] [\fIfile\fR\.\.\.]
.
.SH "DESCRIPTION"
dc(1) is an arbitrary\-precision calculator\. It uses a stack (reverse Polish notation) to store numbers and results of computations\. Arithmetic operations pop arguments off of the stack and push the results\.
.
.P
If no files are given on the command\-line as extra arguments (i\.e\., not as \fB\-f\fR or \fB\-\-file\fR arguments), then dc(1) reads from \fBstdin\fR\. Otherwise, those files are processed, and dc(1) will then exit\.
.
.P
This is different from the dc(1) on OpenBSD and possibly other dc(1) implementations, where \fB\-e\fR (\fB\-\-expression\fR) and \fB\-f\fR (\fB\-\-file\fR) arguments cause dc(1) to execute them and exit\. The reason for this is that this dc(1) allows users to set arguments in the environment variable \fBDC_ENV_ARGS\fR (see the ENVIRONMENT VARIABLES section)\. Any expressions given on the command\-line should be used to set up a standard environment\. For example, if a user wants the \fBscale\fR always set to \fB10\fR, they can set \fBDC_ENV_ARGS\fR to "\-e 10k", and this dc(1) will always start with a \fBscale\fR of \fB10\fR\.
.
.P
If users want to have dc(1) exit after processing all input from \fB\-e\fR and \fB\-f\fR arguments (and their equivalents), then they can just simply add "\-e q" as the last command\-line argument or define the environment variable \fBDC_EXPR_EXIT\fR\.
.
.SH "OPTIONS"
The following are the options that dc(1) accepts\.
.
.TP
\fB\-h\fR, \fB\-\-help\fR
Prints a usage message and quits\.
.
.TP
\fB\-v\fR, \fB\-V\fR, \fB\-\-version\fR
Print the version information (copyright header) and exit\.
.
.TP
\fB\-i\fR, \fB\-\-interactive\fR
Forces interactive mode\. (See the INTERACTIVE MODE section\.)
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.TP
\fB\-P\fR, \fB\-\-no\-prompt\fR
Disables the prompt in interactive mode\. This is mostly for those users that do not want a prompt or are not used to having them in \fBdc\fR\. Most of those users would want to put this option in \fBDC_ENV_ARGS\fR\.
.
.IP
If the prompt has been disabled while building dc(1), this option is a no\-op\.
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.TP
\fB\-x\fR \fB\-\-extended\-register\fR
Enables extended register mode\. See the REGISTERS section for more information\.
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.TP
\fB\-e\fR \fIexpr\fR, \fB\-\-expression\fR=\fIexpr\fR
Evaluates \fBexpr\fR\. If multiple expressions are given, they are evaluated in order\. If files are given as well (see below), the expressions and files are evaluated in the order given\. This means that if a file is given before an expression, the file is read in and evaluated first\.
.
.IP
In other dc(1) implementations, this option causes the program to execute the expressions and then exit\. This dc(1) does not, unless the \fBDC_EXPR_EXIT\fR is defined (see the ENVIRONMENT VARIABLES section)\.
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.TP
\fB\-f\fR \fIfile\fR, \fB\-\-file\fR=\fIfile\fR
Reads in \fBfile\fR and evaluates it\. If expressions are also given (see above), the expressions are evaluated in the order given\.
.
.IP
In other dc(1) implementations, this option causes the program to execute the files and then exit\. This dc(1) does not, unless the \fBDC_EXPR_EXIT\fR is defined (see the ENVIRONMENT VARIABLES section)\.
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.P
\fBNote\fR: long options are only accepted if dc(1) is built with them enabled\.
.
.SH "STDOUT"
Any non\-error output is written to \fBstdout\fR\.
.
.P
\fBNote\fR: Unlike other dc(1) implementations, this dc(1) will issue a fatal error (see the EXIT STATUS section) if it cannot write to \fBstdout\fR, so if \fBstdout\fR is closed, as in \fBdc <file> >&\-\fR, it will quit with an error\. This is done so that dc(1) can report problems when \fBstdout\fR is redirected to a file\.
.
.P
If there are scripts that depend on the behavior of other dc(1) implementations, it is recommended that those scripts be changed to redirect \fBstdout\fR to \fB/dev/null\fR\.
.
.SH "STDERR"
Any error output is written to \fBstderr\fR\.
.
.P
\fBNote\fR: Unlike other dc(1) implementations, this dc(1) will issue a fatal error (see the EXIT STATUS section) if it cannot write to \fBstderr\fR, so if \fBstderr\fR is closed, as in \fBdc <file> 2>&\-\fR, it will quit with an error\. This is done so that dc(1) can report problems when \fBstderr\fR is redirected to a file\.
.
.P
If there are scripts that depend on the behavior of other dc(1) implementations, it is recommended that those scripts be changed to redirect \fBstderr\fR to \fB/dev/null\fR\.
.
.SH "SYNTAX"
\fBibase\fR is a register (see the REGISTERS section) determining how to interpret constant numbers\. It is the "input" base, or the number base used for interpreting input numbers\. \fBibase\fR is initially \fB10\fR\. The max allowable value for \fBibase\fR is \fB16\fR\. The min allowable value for \fBibase\fR is \fB2\fR\. The max allowable value for \fBibase\fR can be queried in dc(1) programs with the \fBT\fR command\.
.
.P
\fBobase\fR is a register (see the REGISTERS section) determining how to output results\. It is the "output" base, or the number base used for outputting numbers\. \fBobase\fR is initially \fB10\fR\. The max allowable value for \fBobase\fR is \fBDC_BASE_MAX\fR\. The min allowable value for \fBobase\fR is \fB2\fR unless dc(1) was built with the extra math option\. If it was, then the min allowable value is \fB0\fR\. In this case, if \fBobase\fR is \fB0\fR, values are output in scientific notation, and if \fBobase\fR is \fB1\fR, values are output in engineering notation\. (Outputting in scientific or engineering notation are \fBnon\-portable extensions\fR\.) The max allowable value for \fBobase\fR can be queried in dc(1) programs with the \fBU\fR command\.
.
.P
The \fBscale\fR of an expression is the number of digits in the result of the expression right of the decimal point, and \fBscale\fR is a register (see the REGISTERS section) that sets the precision of any operations (with exceptions)\. \fBscale\fR is initially \fB0\fR\. \fBscale\fR cannot be negative\. The max allowable value for \fBscale\fR can be queried in dc(1) programs with the \fBV\fR command\.
.
.P
Each item in the input source code, either a number (see the NUMBERS section) or a command (see the COMMANDS section), is processed and executed, in order\. Input is processed immediately when entered\.
.
.P
If dc(1) was built with the extra math option, there is an additional register named \fBseed\fR\. This is the current seed used by the pseudo\-random number generator\. If the current value of \fBseed\fR is queried and stored, then if it is assigned to \fBseed\fR later, the pseudo\-random number generator is guaranteed to produce the same sequence of pseudo\-random numbers that were generated after the value of \fBseed\fR was first queried\.
.
.P
Multiple values assigned to \fBseed\fR can produce the same sequence of pseudo\-random numbers\. Likewise, when a value is assigned to \fBseed\fR, it is not guaranteed that querying \fBseed\fR immediately after will return the same value\. In addition, the value of \fBseed\fR will change after any call to the \fB'\fR or \fB"\fR commands\. The maximum integer returned by the \fB'\fR command can be queried with the \fBW\fR command\.
.
.P
\fBNote\fR: The values returned by the pseudo\-random number generator with the \fB'\fR and \fB"\fR commands are guaranteed to \fBNOT\fR be cryptographically\-secure\. This is a consequence of using a seeded pseudo\-random number generator\. However, they \fBare\fR guaranteed to be reproducible with identical \fBseed\fR values\.
.
.P
The pseudo\-random number generator, \fBseed\fR, and all associated operations are \fBnon\-portable extensions\fR\.
.
.SS "Comments"
Comments go from \fB#\fR until, and not including, the next newline\. This is a \fBnon\-portable extension\fR\.
.
.SH "NUMBERS"
Numbers are strings made up of digits, uppercase letters up to \fBF\fR, and at most \fB1\fR period for a radix\. Numbers can have up to \fBDC_NUM_MAX\fR digits\. Uppercase letters equal \fB9\fR + their position in the alphabet (i\.e\., \fBA\fR equals \fB10\fR, or \fB9 + 1\fR)\. If a digit or letter makes no sense with the current value of \fBibase\fR, they are set to the value of the highest valid digit in \fBibase\fR\.
.
.P
Single\-character numbers (i\.e\., \fBA\fR) take the value that they would have if they were valid digits, regardless of the value of \fBibase\fR\. This means that \fBA\fR always equals decimal \fB10\fR and \fBF\fR always equals decimal \fB15\fR\.
.
.P
In addition, if dc(1) was built with the extra math option, it accepts numbers in scientific notation\. For dc(1), an example is \fB1\.89237e9\fR, which is equal to \fB1892370000\fR\. Negative exponents are also allowed, so \fB4\.2890e_3\fR is equal to \fB0\.0042890\fR\.
.
.P
\fBWARNING\fR: Both the number and the exponent in scientific notation are interpreted according to the current \fBibase\fR, but the number is still multiplied by \fB10^exponent\fR regardless of the current \fBibase\fR\. For example, if \fBibase\fR is \fB16\fR and dc(1) is given the number string \fB"FFeA"\fR, the resulting decimal number will be \fB2550000000000\fR, and if dc(1) is given the number string \fB"10e_4"\fR, the resulting decimal number will be \fB0\.0016\fR\.
.
.P
Accepting input as scientific notation is a \fBnon\-portable extension\fR\.
.
.SH "COMMANDS"
The valid commands are listed below\.
.
.SS "Printing"
These commands are used for printing\.
.
.P
Note that if dc(1) has been built with the extra math option enabled, both scientific notation and engineering notation are available for printing numbers\. Scientific notation is activated by assigning \fB0\fR to \fBobase\fR using \fB0o\fR (in any other context, an \fBobase\fR of \fB0\fR is invalid), and engineering notation is activated by assigning \fB1\fR to \fBobase\fR using \fB1o\fR (which is also invalid in any other context)\. To deactivate them, just assign a different value to \fBobase\fR\.
.
.P
Printing numbers in scientific notation and/or engineering notation is a \fBnon\-portable extension\fR\.
.
.TP
\fBp\fR
Prints the value on top of the stack, whether number or string, and prints a newline after\.
.
.IP
This does not alter the stack\.
.
.TP
\fBn\fR
Prints the value on top of the stack, whether number or string, and pops it off of the stack\.
.
.TP
\fBP\fR
Pops a value off the stack\.
.
.IP
If the value is a number, it is truncated and the absolute value of the result is printed as though \fBobase\fR is \fBUCHAR_MAX + 1\fR and each digit is interpreted as an ASCII character, making it a byte stream\.
.
.IP
If the value is a string, it is printed without a trailing newline\.
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.TP
\fBf\fR
Prints the entire contents of the stack, in order from newest to oldest, without altering anything\.
.
.IP
Users should use this command when they get lost\.
.
.SS "Arithmetic"
These are the commands used for arithmetic\.
.
.TP
\fB+\fR
The top two values are popped off the stack, added, and the result is pushed onto the stack\. The \fBscale\fR of the result is equal to the max \fBscale\fR of both operands\.
.
.TP
\fB\-\fR
The top two values are popped off the stack, subtracted, and the result is pushed onto the stack\. The \fBscale\fR of the result is equal to the max \fBscale\fR of both operands\.
.
.TP
\fB*\fR
The top two values are popped off the stack, multiplied, and the result is pushed onto the stack\. If \fBa\fR is the \fBscale\fR of the first expression and \fBb\fR is the \fBscale\fR of the second expression, the \fBscale\fR of the result is equal to \fBmin(a+b,max(scale,a,b))\fR where \fBmin\fR and \fBmax\fR return the obvious values\.
.
.TP
\fB/\fR
The top two values are popped off the stack, divided, and the result is pushed onto the stack\. The \fBscale\fR of the result is equal to \fBscale\fR\.
.
.IP
The first value popped off of the stack must be non\-zero\.
.
.TP
\fB%\fR
The top two values are popped off the stack, remaindered, and the result is pushed onto the stack\.
.
.IP
Remaindering is equivalent to 1) Computing \fBa/b\fR to current \fBscale\fR, and 2) Using the result of step 1 to calculate \fBa\-(a/b)*b\fR to \fBscale\fR \fBmax(scale + scale(b), scale(a))\fR\.
.
.IP
The first value popped off of the stack must be non\-zero\.
.
.TP
\fB~\fR
The top two values are popped off the stack, divided and remaindered, and the results (divided first, remainder second) are pushed onto the stack\. This is equivalent to \fBx y / x y %\fR except that \fBx\fR and \fBy\fR are only evaluated once\.
.
.IP
The first value popped off of the stack must be non\-zero\.
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.TP
\fB^\fR
The top two values are popped off the stack, the second is raised to the power of the first, and the result is pushed onto the stack\.
.
.IP
The first value popped off of the stack must be an integer, and if that value is negative, the second value popped off of the stack must be non\-zero\.
.
.TP
\fBv\fR
The top value is popped off the stack, its square root is computed, and the result is pushed onto the stack\. The \fBscale\fR of the result is equal to \fBscale\fR\.
.
.IP
The value popped off of the stack must be non\-negative\.
.
.TP
\fB_\fR
If this command \fIimmediately\fR precedes a number (i\.e\., no spaces or other commands), then that number is input as a negative number\.
.
.IP
Otherwise, the top value on the stack is popped and copied, and the copy is negated and pushed onto the stack\. This behavior without a number is a \fBnon\-portable extension\fR\.
.
.TP
\fBb\fR
The top value is popped off the stack, and if it is zero, it is pushed back onto the stack\. Otherwise, its absolute value is pushed onto the stack\.
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.TP
\fB|\fR
The top three values are popped off the stack, a modular exponentiation is computed, and the result is pushed onto the stack\.
.
.IP
The first value popped is used as the reduction modulus and must be an integer and non\-zero\. The second value popped is used as the exponent and must be an integer and non\-negative\. The third value popped is the base and must be an integer\.
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.TP
\fB$\fR
The top value is popped off the stack and copied, and the copy is truncated and pushed onto the stack\.
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.TP
\fB@\fR
The top two values are popped off the stack, and the precision of the second is set to the value of the first, whether by truncation or extension\.
.
.IP
The first value popped off of the stack must be an integer and non\-negative\.
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.TP
\fBH\fR
The top two values are popped off the stack, and the second is shifted left (radix shifted right) to the value of the first\.
.
.IP
The first value popped off of the stack must be an integer and non\-negative\.
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.TP
\fBh\fR
The top two values are popped off the stack, and the second is shifted right (radix shifted left) to the value of the first\.
.
.IP
The first value popped off of the stack must be an integer and non\-negative\.
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.TP
\fBG\fR
The top two values are popped off of the stack, they are compared, and a \fB1\fR is pushed if they are equal, or \fB0\fR otherwise\.
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.TP
\fBN\fR
The top value is popped off of the stack, and if it a \fB0\fR, a \fB1\fR is pushed; otherwise, a \fB0\fR is pushed\.
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.TP
\fB(\fR
The top two values are popped off of the stack, they are compared, and a \fB1\fR is pushed if the first is less than the second, or \fB0\fR otherwise\.
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.TP
\fB{\fR
The top two values are popped off of the stack, they are compared, and a \fB1\fR is pushed if the first is less than or equal to the second, or \fB0\fR otherwise\.
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.TP
\fB)\fR
The top two values are popped off of the stack, they are compared, and a \fB1\fR is pushed if the first is greater than the second, or \fB0\fR otherwise\.
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.TP
\fB}\fR
The top two values are popped off of the stack, they are compared, and a \fB1\fR is pushed if the first is greater than or equal to the second, or \fB0\fR otherwise\.
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.TP
\fBM\fR
The top two values are popped off of the stack\. If they are both non\-zero, a \fB1\fR is pushed onto the stack\. If either of them is zero, or both of them are, then a \fB0\fR is pushed onto the stack\.
.
.IP
This is like the \fB&&\fR operator in bc(1), and it is not a short\-circuit operator\.
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.TP
\fBm\fR
The top two values are popped off of the stack\. If at least one of them is non\-zero, a \fB1\fR is pushed onto the stack\. If both of them are zero, then a \fB0\fR is pushed onto the stack\.
.
.IP
This is like the \fB||\fR operator in bc(1), and it is not a short\-circuit operator\.
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.SS "Pseudo\-Random Number Generator"
If dc(1) was built with the extra math option, it has a built\-in pseudo\-random number generator\. These commands query the pseudo\-random number generator\. (See Parameters for more information about the \fBseed\fR value that controls the pseudo\-random number generator\.)
.
.P
The pseudo\-random number generator is guaranteed to \fBNOT\fR be cryptographically\-secure\.
.
.TP
\fB'\fR
Generates an integer between 0 and \fBDC_RAND_MAX\fR, inclusive (see the LIMITS section)\.
.
.IP
The generated integer is made as unbiased as possible, subject to the limitations of the pseudo\-random number generator\.
.
.IP
This command is only available if dc(1) was built with the extra math option\.
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.TP
\fB"\fR
Pops a value off of the stack, which is used as an \fBexclusive\fR upper bound on the integer that will be generated\. If the bound is negative or is a non\-integer, an error is raised, and dc(1) resets (see the RESET section)\. If the bound is larger than \fBDC_RAND_MAX\fR, the higher bound is honored by generating several pseudo\-random integers, multiplying them by appropriate powers of \fBDC_RAND_MAX + 1\fR, and adding them together\. Thus, the size of integer that can be generated with this command is unbounded\. Using this command will change the value of \fBseed\fR\.
.
.IP
If the operand is \fB0\fR or \fB1\fR, then the result pushed onto the stack is \fB0\fR, and \fBseed\fR is not changed\.
.
.IP
The generated integer is made as unbiased as possible, subject to the limitations of the pseudo\-random number generator\.
.
.IP
This command is only available if dc(1) was built with the extra math option\.
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.SS "Stack Control"
These commands control the stack\.
.
.TP
\fBc\fR
Removes all items from ("clears") the stack\.
.
.TP
\fBd\fR
Copies the item on top of the stack ("duplicates") and pushes the copy onto the stack\.
.
.TP
\fBr\fR
Swaps ("reverses") the two top items on the stack\.
.
.TP
\fBR\fR
Pops ("removes") the top value from the stack\.
.
.SS "Register Control"
These commands control registers (see the REGISTERS section)\.
.
.TP
\fBs\fR\fIr\fR
Pops the value off the top of the stack and stores it into register \fBr\fR\.
.
.TP
\fBl\fR\fIr\fR
Copies the value in register \fBr\fR and pushes it onto the stack\. This does not alter the contents of \fBr\fR\.
.
.TP
\fBS\fR\fIr\fR
Pops the value off the top of the (main) stack and pushes it onto the stack of register \fBr\fR\. The previous value of the register becomes inaccessible\.
.
.TP
\fBL\fR\fIr\fR
Pops the value off the top of the stack for register \fBr\fR and push it onto the main stack\. The previous value in the stack for register \fBr\fR, if any, is now accessible via the \fBl\fR\fIr\fR command\.
.
.SS "Parameters"
These commands control the values of \fBibase\fR, \fBobase\fR, \fBscale\fR, and \fBseed\fR (if dc(1) was built with the extra math option)\. Also see the SYNTAX section\.
.
.TP
\fBi\fR
Pops the value off of the top of the stack and uses it to set \fBibase\fR, which must be between \fB2\fR and \fB16\fR, inclusive\.
.
.IP
If the value on top of the stack has any \fBscale\fR, the \fBscale\fR is ignored\.
.
.TP
\fBo\fR
Pops the value off of the top of the stack and uses it to set \fBobase\fR, which must be between \fB2\fR and \fBDC_BASE_MAX\fR, inclusive (see bc(1))\. The value can be either \fB0\fR or \fB1\fR if dc(1) was built with the extra math option\.
.
.IP
If the value on top of the stack has any \fBscale\fR, the \fBscale\fR is ignored\.
.
.TP
\fBk\fR
Pops the value off of the top of the stack and uses it to set \fBscale\fR, which must be non\-negative\.
.
.IP
If the value on top of the stack has any \fBscale\fR, the \fBscale\fR is ignored\.
.
.TP
\fBj\fR
Pops the value off of the top of the stack and uses it to set \fBseed\fR\. The meaning of \fBseed\fR is dependent on the current pseudo\-random number generator but is guaranteed to not change except for new major versions\.
.
.IP
The \fBscale\fR of the value may be significant\.
.
.IP
If a previously used \fBseed\fR value is used again, the pseudo\-random number generator is guaranteed to produce the same sequence of pseudo\-random numbers as it did when the \fBseed\fR value was previously used\.
.
.IP
The exact value assigned to \fBseed\fR is not guaranteed to be returned if the \fBJ\fR command is used\. However, if \fBseed\fR \fIdoes\fR return a different value, both values, when assigned to \fBseed\fR, are guaranteed to produce the same sequence of pseudo\-random numbers\. This means that certain values assigned to \fBseed\fR will not produce unique sequences of pseudo\-random numbers\.
.
.IP
There is no limit to the length (number of significant decimal digits) or \fIscale\fR of the value that can be assigned to \fBseed\fR\.
.
.IP
This command is only available if dc(1) was built with the extra math option\.
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.TP
\fBI\fR
Pushes the current value of \fBibase\fR onto the main stack\.
.
.TP
\fBO\fR
Pushes the current value of \fBobase\fR onto the main stack\.
.
.TP
\fBK\fR
Pushes the current value of \fBscale\fR onto the main stack\.
.
.TP
\fBJ\fR
Pushes the current value of \fBseed\fR onto the main stack\.
.
.IP
This command is only available if dc(1) was built with the extra math option\.
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.TP
\fBT\fR
Pushes the maximum allowable value of \fBibase\fR onto the main stack\.
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.TP
\fBU\fR
Pushes the maximum allowable value of \fBobase\fR onto the main stack\.
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.TP
\fBV\fR
Pushes the maximum allowable value of \fBscale\fR onto the main stack\.
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.TP
\fBW\fR
Pushes the maximum (inclusive) integer that can be generated with the \fB'\fR pseudo\-random number generator command\.
.
.IP
This command is only available if dc(1) was built with the extra math option\.
.
.IP
This is a \fBnon\-portable extension\fR\.
.
.SS "Strings"
The following commands control strings\.
.
.P
dc(1) can work with both numbers and strings, and registers (see the REGISTERS section) can hold both strings and numbers\. dc(1) always knows whether the contents of a register are a string or a number\.
.
.P
While arithmetic operations have to have numbers, and will print an error if given a string, other commands accept strings\.
.
.P
Strings can also be executed as macros\. For example, if the string \fB[1pR]\fR is executed as a macro, then the code \fB1pR\fR is executed, meaning that the \fB1\fR will be printed with a newline after and then popped from the stack\.
.
.TP
\fB[\fR\fIcharacters\fR\fB]\fR
Makes a string containing \fIcharacters\fR and pushes it onto the stack\.
.
.IP
If there are brackets (\fB[\fR and \fB]\fR) in the string, then they must be balanced\. Unbalanced brackets can be escaped using a backslash (\fB\e\fR) character\.
.
.IP

NEWS
--
NTP 4.2.8p12 (Harlan Stenn <stenn@ntp.org>, 2018/14/09)

NOTE: this NEWS file will be undergoing more revisions.

Focus: Security, Bug fixes, enhancements.

Severity: MEDIUM

This release fixes a "hole" in the noepeer capability introduced to ntpd
in ntp-4.2.8p11, and a buffer overflow in the openhost() function used by
ntpq and ntpdc.  It also provides 26 other bugfixes, and 4 other improvements:

* [Sec 3505] Buffer overflow in the openhost() call of ntpq and ntpdc.

* [Sec 3012] Fix a hole in the new "noepeer" processing.

* Bug Fixes:
 [Bug 3521] Fix a logic bug in the INVALIDNAK checks.  <stenn@ntp.org>
 [Bug 3509] Add support for running as non-root on FreeBSD, Darwin,
            other TrustedBSD platforms
 - applied patch by Ian Lepore <perlinger@ntp.org>
 [Bug 3506] Service Control Manager interacts poorly with NTPD <perlinger@ntp.org>
 - changed interaction with SCM to signal pending startup
 [Bug 3486] Buffer overflow in ntpq/ntpq.c:tstflags() <perlinger@ntp.org>
 - applied patch by Gerry Garvey
 [Bug 3485] Undefined sockaddr used in error messages in ntp_config.c <perlinger@ntp.org>
 - applied patch by Gerry Garvey
 [Bug 3484] ntpq response from ntpd is incorrect when REFID is null <perlinger@ntp.org>
 - rework of ntpq 'nextvar()' key/value parsing
 [Bug 3482] Fixes for compilation warnings (ntp_io.c & ntpq-subs.c) <perlinger@ntp.org>
 - applied patch by Gerry Garvey (with mods)
 [Bug 3480] Refclock sample filter not cleared on clock STEP <perlinger@ntp.org>
 - applied patch by Gerry Garvey
 [Bug 3479] ctl_putrefid() allows unsafe characters through to ntpq <perlinger@ntp.org>
 - applied patch by Gerry Garvey (with mods)
 [Bug 3476]ctl_putstr() sends empty unquoted string [...] <perlinger@ntp.org>
 - applied patch by Gerry Garvey (with mods); not sure if that's bug or feature, though
 [Bug 3475] modify prettydate() to suppress output of zero time <perlinger@ntp.org>
 - applied patch by Gerry Garvey
 [Bug 3474] Missing pmode in mode7 peer info response <perlinger@ntp.org>
 - applied patch by Gerry Garvey
 [Bug 3471] Check for openssl/[ch]mac.h.  HStenn.
 - add #define ENABLE_CMAC support in configure.  HStenn.
 [Bug 3470] ntpd4.2.8p11 fails to compile without OpenSSL <perlinger@ntp.org>
 [Bug 3469] Incomplete string compare [...] in is_refclk_addr <perlinger@ntp.org>
 - patch by Stephen Friedl
 [Bug 3467] Potential memory fault in ntpq [...] <perlinger@ntp.org>
 - fixed IO redirection and CTRL-C handling in ntq and ntpdc
 [Bug 3465] Default TTL values cannot be used <perlinger@ntp.org>
 [Bug 3461] refclock_shm.c: clear error status on clock recovery <perlinger@ntp.org>
 - initial patch by Hal Murray; also fixed refclock_report() trouble
 [Bug 3460] Fix typo in ntpq.texi, reported by Kenyon Ralph.  <stenn@ntp.org>
 [Bug 3456] Use uintptr_t rather than size_t to store an integer in a pointer
 - According to Brooks Davis, there was only one location <perlinger@ntp.org>
 [Bug 3449] ntpq - display "loop" instead of refid [...] <perlinger@ntp.org>
 - applied patch by Gerry Garvey
 [Bug 3445] Symmetric peer won't sync on startup <perlinger@ntp.org>
 - applied patch by Gerry Garvey
 [Bug 3442] Fixes for ntpdate as suggested by Gerry Garvey,
 with modifications
 New macro REFID_ISTEXT() which is also used in ntpd/ntp_control.c.
 [Bug 3434] ntpd clears STA_UNSYNC on start <perlinger@ntp.org>
 - applied patch by Miroslav Lichvar
 [Bug 3426] ntpdate.html -t default is 2 seconds.  Leonid Evdokimov.
 [Bug 3121] Drop root privileges for the forked DNS worker <perlinger@ntp.org>
 - integrated patch by  Reinhard Max
 [Bug 2821] minor build issues <perlinger@ntp.org>
 - applied patches by Christos Zoulas, including real bug fixes
 html/authopt.html: cleanup, from <stenn@ntp.org>
 ntpd/ntpd.c: DROPROOT cleanup.  <stenn@ntp.org>
 Symmetric key range is 1-65535.  Update docs.   <stenn@ntp.org>

--
NTP 4.2.8p11 (Harlan Stenn <stenn@ntp.org>, 2018/02/27)

Focus: Security, Bug fixes, enhancements.

Severity: MEDIUM

This release fixes 2 low-/medium-, 1 informational/medum-, and 2 low-severity
vulnerabilities in ntpd, one medium-severity vulernability in ntpq, and
provides 65 other non-security fixes and improvements:

* NTP Bug 3454: Unauthenticated packet can reset authenticated interleaved
	association (LOW/MED)
   Date Resolved: Stable (4.2.8p11) 27 Feb 2018
   References: Sec 3454 / CVE-2018-7185 / VU#961909
   Affects: ntp-4.2.6, up to but not including ntp-4.2.8p11.
   CVSS2: MED 4.3 (AV:N/AC:M/Au:N/C:N/I:N/A:P) This could score between
	2.9 and 6.8.
   CVSS3: LOW 3.1 CVSS:3.0/AV:N/AC:H/PR:L/UI:N/S:U/C:N/I:N/A:L This could
	score between 2.6 and 3.1
   Summary:
	The NTP Protocol allows for both non-authenticated and
	authenticated associations, in client/server, symmetric (peer),
	and several broadcast modes. In addition to the basic NTP
	operational modes, symmetric mode and broadcast servers can
	support an interleaved mode of operation. In ntp-4.2.8p4 a bug
	was inadvertently introduced into the protocol engine that
	allows a non-authenticated zero-origin (reset) packet to reset
	an authenticated interleaved peer association. If an attacker
	can send a packet with a zero-origin timestamp and the source
	IP address of the "other side" of an interleaved association,
	the 'victim' ntpd will reset its association. The attacker must
	continue sending these packets in order to maintain the
	disruption of the association. In ntp-4.0.0 thru ntp-4.2.8p6,
	interleave mode could be entered dynamically. As of ntp-4.2.8p7,
	interleaved mode must be explicitly configured/enabled.
   Mitigation:
	Implement BCP-38.
	Upgrade to 4.2.8p11, or later, from the NTP Project Download Page
	    or the NTP Public Services Project Download Page.
	If you are unable to upgrade to 4.2.8p11 or later and have
	    'peer HOST xleave' lines in your ntp.conf file, remove the
	    'xleave' option.
	Have enough sources of time.
	Properly monitor your ntpd instances.
	If ntpd stops running, auto-restart it without -g .

src/dc/dc.c
/*
 * *****************************************************************************
 *
 * Copyright (c) 2018-2020 Gavin D. Howard and contributors.
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * *****************************************************************************
 *
 * The main procedure of dc.
 *
 */

#if DC_ENABLED

#include <string.h>

#include <status.h>
#include <dc.h>
#include <vm.h>

void dc_main(int argc, char **argv) {

	vm.read_ret = BC_INST_POP_EXEC;
	vm.help = dc_help;
	vm.sigmsg = dc_sig_msg;
	vm.siglen = dc_sig_msg_len;

	vm.next = dc_lex_token;
	vm.parse = dc_parse_parse;
	vm.expr = dc_parse_expr;

	bc_vm_boot(argc, argv, "DC_LINE_LENGTH", "DC_ENV_ARGS", "DC_EXPR_EXIT");
}
#endif // DC_ENABLED
